File: Shared/Routine.swift
import Foundation

enum RoutineType: String, Codable, CaseIterable, Sendable {
    case check
    case count
}

enum RepeatMode: String, Codable, CaseIterable, Sendable {
    case daily     // 매일
    case weekdays  // 주간 (평일 월-금)
    case custom    // 직접 선택

    var label: String {
        switch self {
        case .daily:    return "매일"
        case .weekdays: return "주간"
        case .custom:   return "직접 선택"
        }
    }

    var defaultDays: [Int] {
        switch self {
        case .daily:    return Array(0...6)          // 모든 요일
        case .weekdays: return [1, 2, 3, 4, 5]       // 월~금
        case .custom:   return []
        }
    }
}

struct Routine: Identifiable, Codable, Equatable, Sendable {
    var id: UUID
    var name: String
    var type: RoutineType
    var target: Int
    var unit: String
    var reminder: String?   // "HH:MM" or nil
    var anytime: Bool
    var repeatMode: RepeatMode
    var repeatDays: [Int]   // 0=일 1=월 … 6=토 (custom 모드에서 사용)
    var createdAt: Date     // 시작일: 이 날짜(당일 포함)부터 루틴이 노출된다
    var endDate: Date?      // 종료일: 이 날짜(당일 포함)부터 루틴이 사라진다 (nil = 무기한). 이전 기록은 보존.

    // 해당 요일에 이 루틴이 예정됐는지 (weekday: 0=일)
    func isScheduled(on weekday: Int) -> Bool {
        switch repeatMode {
        case .daily:    return true
        case .weekdays: return (1...5).contains(weekday)
        case .custom:   return repeatDays.contains(weekday)
        }
    }

    // 해당 날짜에 이 루틴이 예정됐는지. 생성일 이전, 종료일 당일/이후 날짜에는 표시하지 않는다.
    func isScheduled(on date: Date) -> Bool {
        let cal = Calendar.gregorianSunday
        let targetDay = cal.startOfDay(for: date)
        let createdDay = cal.startOfDay(for: createdAt)
        guard targetDay >= createdDay else { return false }
        if let endDate, targetDay >= cal.startOfDay(for: endDate) { return false }

        let weekday = cal.component(.weekday, from: targetDay) - 1
        return isScheduled(on: weekday)
    }

    init(
        id: UUID = UUID(),
        name: String,
        type: RoutineType = .check,
        target: Int = 1,
        unit: String = "",
        reminder: String? = nil,
        anytime: Bool = true,
        repeatMode: RepeatMode = .daily,
        repeatDays: [Int] = [],
        createdAt: Date = Date(),
        endDate: Date? = nil
    ) {
        self.id = id
        self.name = name
        self.type = type
        self.target = target
        self.unit = unit
        self.reminder = reminder
        self.anytime = anytime
        self.repeatMode = repeatMode
        self.repeatDays = repeatDays
        self.createdAt = createdAt
        self.endDate = endDate
    }

    // 이전 버전 데이터(repeatMode/repeatDays/createdAt 없음) 호환 디코더
    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        id         = try c.decode(UUID.self,        forKey: .id)
        name       = try c.decode(String.self,      forKey: .name)
        type       = try c.decode(RoutineType.self, forKey: .type)
        target     = try c.decode(Int.self,         forKey: .target)
        unit       = try c.decode(String.self,      forKey: .unit)
        reminder   = try c.decodeIfPresent(String.self,     forKey: .reminder)
        anytime    = try c.decode(Bool.self,        forKey: .anytime)
        repeatMode = try c.decodeIfPresent(RepeatMode.self, forKey: .repeatMode) ?? .daily
        repeatDays = try c.decodeIfPresent([Int].self,      forKey: .repeatDays) ?? []
        createdAt  = try c.decodeIfPresent(Date.self,        forKey: .createdAt) ?? .distantPast
        endDate    = try c.decodeIfPresent(Date.self,        forKey: .endDate)
    }
}

File: server/src/main/java/com/routinecalendar/server/routine/domain/Routine.java
package com.routinecalendar.server.routine.domain;

import com.routinecalendar.server.user.domain.User;
import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.FetchType;
import jakarta.persistence.Id;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.PostLoad;
import jakarta.persistence.PostPersist;
import jakarta.persistence.Table;
import jakarta.persistence.Transient;
import java.time.Instant;
import java.time.LocalDate;
import java.util.ArrayList;
import java.util.List;
import java.util.UUID;
import lombok.AccessLevel;
import lombok.Getter;
import lombok.NoArgsConstructor;
import org.hibernate.annotations.JdbcTypeCode;
import org.hibernate.annotations.UpdateTimestamp;
import org.hibernate.type.SqlTypes;
import org.springframework.data.domain.Persistable;

/**
 * 루틴 원본. id는 클라이언트(iOS)가 만든 UUID를 그대로 쓴다(기기 간 동일 식별).
 * 삭제는 {@code deletedAt} soft delete — 다른 기기가 삭제를 인지할 수 있게 한다.
 */
@Entity
@Table(name = "routines")
@Getter
@NoArgsConstructor(access = AccessLevel.PROTECTED)
public class Routine implements Persistable<UUID> {

    @Id
    private UUID id;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "user_id", nullable = false)
    private User user;

    @Column(nullable = false, length = 100)
    private String name;

    @Column(nullable = false, length = 10)
    private String type;        // check / count

    @Column(nullable = false)
    private int target;

    @Column(nullable = false, length = 20)
    private String unit;

    @Column(length = 5)
    private String reminder;    // "HH:MM" or null

    @Column(nullable = false)
    private boolean anytime;

    @Column(name = "repeat_mode", nullable = false, length = 10)
    private String repeatMode;  // daily / weekdays / custom

    @JdbcTypeCode(SqlTypes.JSON)
    @Column(name = "repeat_days", nullable = false)
    private List<Integer> repeatDays = new ArrayList<>();

    // 시작일: 클라(iOS)가 캘린더에서 선택한 날짜. 이 날짜(당일 포함)부터 루틴이 노출된다.
    // 자동 타임스탬프(@CreationTimestamp)를 쓰지 않고 클라가 보낸 값을 그대로 영속화한다.
    @Column(name = "created_at", nullable = false, updatable = false)
    private Instant createdAt;

    // 종료일: 이 날짜(당일 포함)부터 루틴이 사라진다(nil = 무기한). 이전 기록은 보존.
    @Column(name = "end_date")
    private LocalDate endDate;

    @UpdateTimestamp
    @Column(name = "updated_at", nullable = false)
    private Instant updatedAt;

    @Column(name = "deleted_at")
    private Instant deletedAt;

    /** 서버 푸시 리마인더로 오늘 평가/발송한 날짜 (하루 1회 보장). */
    @Column(name = "last_reminded_on")
    private LocalDate lastRemindedOn;

    // 할당된 UUID PK라 save() 시 insert/update를 구분하기 위한 플래그.
    @Transient
    private boolean isNew = true;

    public Routine(UUID id, User user, String name, String type, int target, String unit,
                   String reminder, boolean anytime, String repeatMode, List<Integer> repeatDays,
                   Instant createdAt, LocalDate endDate) {
        this.id = id;
        this.user = user;
        this.name = name;
        this.type = type;
        this.target = target;
        this.unit = unit;
        this.reminder = reminder;
        this.anytime = anytime;
        this.repeatMode = repeatMode;
        this.repeatDays = repeatDays != null ? repeatDays : new ArrayList<>();
        this.createdAt = createdAt != null ? createdAt : Instant.now();
        this.endDate = endDate;
    }

    public void update(String name, String type, int target, String unit, String reminder,
                       boolean anytime, String repeatMode, List<Integer> repeatDays,
                       LocalDate endDate) {
        this.name = name;
        this.type = type;
        this.target = target;
        this.unit = unit;
        this.reminder = reminder;
        this.anytime = anytime;
        this.repeatMode = repeatMode;
        this.repeatDays = repeatDays != null ? repeatDays : new ArrayList<>();
        this.endDate = endDate;
    }

    public void markDeleted() {
        this.deletedAt = Instant.now();
    }

    /** 오늘 리마인더를 처리했다고 기록(중복 발송 방지). */
    public void markReminded(LocalDate date) {
        this.lastRemindedOn = date;
    }

    /** 이 루틴이 해당 요일(0=일 … 6=토)에 예정돼 있는지. (iOS isScheduled와 동일 규칙) */
    public boolean isScheduledOn(int weekday) {
        return switch (repeatMode) {
            case "weekdays" -> weekday >= 1 && weekday <= 5;
            case "custom"   -> repeatDays.contains(weekday);
            default          -> true;   // daily
        };
    }

    @Override
    public boolean isNew() {
        return isNew;
    }

    @PostLoad
    @PostPersist
    void markNotNew() {
        this.isNew = false;
    }
}
